In [149]:
library(class)
library(ggplot2)

In [167]:
german<- read.csv("/Users/vanshika/Desktop/Data Intensive Computing/lab3Data-selected/german_credit_data1.csv")

In [168]:
german

X.  Age Sex    Job Housing Saving.accounts Checking.account Credit.amount
1     0  67  male   2   own     NA              little            1169        
2     1  22  female 2   own     little          moderate          5951        
3     2  49  male   1   own     little          NA                2096        
4     3  45  male   2   free    little          little            7882        
5     4  53  male   2   free    little          little            4870        
6     5  35  male   1   free    NA              NA                9055        
7     6  53  male   2   own     quite rich      NA                2835        
8     7  35  male   3   rent    little          moderate          6948        
9     8  61  male   1   own     rich            NA                3059        
10    9  28  male   3   own     little          moderate          5234        
11   10  25  female 2   rent    little          moderate          1295        
12   11  24  female 2   rent    little          little            4308        
13   12  22  female 2   own     little          moderate          1567        
14   13  60  male   1   own     little          little            1199        
15   14  28  female 2   rent    little          little            1403        
16   15  32  female 1   own     moderate        little            1282        
17   16  53  male   2   own     NA              NA                2424        
18   17  25  male   2   own     NA              little            8072        
19   18  44  female 3   free    little          moderate         12579        
20   19  31  male   2   own     quite rich      NA                3430        
21   20  48  male   2   own     little          NA                2134        
22   21  44  male   2   rent    quite rich      little            2647        
23   22  48  male   1   rent    little          little            2241        
24   23  44  male   2   own     moderate        moderate          1804        
25   24  26  male   2   own     NA              NA                2069        
26   25  36  male   1   own     little          little            1374        
27   26  39  male   1   own     little          NA                 426        
28   27  42  female 2   rent    rich            rich               409        
29   28  34  male   2   own     little          moderate          2415        
30   29  63  male   2   own     little          little            6836        
⋮    ⋮   ⋮   ⋮      ⋮   ⋮       ⋮               ⋮                ⋮            
971  970 22  male   2   own     moderate        moderate         1514         
972  971 43  male   1   own     little          NA               7393         
973  972 29  female 0   rent    little          little           1193         
974  973 36  male   2   rent    little          little           7297         
975  974 33  female 2   own     little          NA               2831         
976  975 57  female 1   own     quite rich      rich             1258         
977  976 64  female 2   own     little          moderate          753         
978  977 42  male   2   own     NA              moderate         2427         
979  978 47  male   1   own     little          NA               2538         
980  979 25  male   2   rent    moderate        moderate         1264         
981  980 49  male   2   own     little          moderate         8386         
982  981 33  male   3   rent    little          NA               4844         
983  982 28  female 3   own     moderate        rich             2923         
984  983 26  male   2   own     little          little           8229         
985  984 30  male   1   own     little          NA               2028         
986  985 25  female 2   rent    little          little           1433         
987  986 33  male   2   own     little          rich             6289         
988  987 64  female 2   own     moderate        NA               1409         
989  988 29  male   3   free    little          little 

In [152]:
str(german)

'data.frame':	1000 obs. of  11 variables:
 $ X.              : int  0 1 2 3 4 5 6 7 8 9 ...
 $ Age             : int  67 22 49 45 53 35 53 35 61 28 ...
 $ Sex             : Factor w/ 2 levels "female","male": 2 1 2 2 2 2 2 2 2 2 ...
 $ Job             : int  2 2 1 2 2 1 2 3 1 3 ...
 $ Housing         : Factor w/ 3 levels "free","own","rent": 2 2 2 1 1 1 2 3 2 2 ...
 $ Saving.accounts : Factor w/ 4 levels "little","moderate",..: NA 1 1 1 1 NA 3 1 4 1 ...
 $ Checking.account: Factor w/ 3 levels "little","moderate",..: 1 2 NA 1 1 NA NA 2 NA 2 ...
 $ Credit.amount   : int  1169 5951 2096 7882 4870 9055 2835 6948 3059 5234 ...
 $ Duration        : int  6 48 12 42 24 36 24 36 12 30 ...
 $ Purpose         : Factor w/ 8 levels "business","car",..: 6 6 4 5 2 4 5 2 6 2 ...
 $ Credit.Risks    : int  1 2 1 1 2 1 1 1 1 2 ...


In [169]:
german_data<-sapply(german, as.numeric)
german_data

X.  Age Sex Job Housing Saving.accounts Checking.account Credit.amount
 [1,]  0  67  2   2   2       NA               1                1169        
 [2,]  1  22  1   2   2        1               2                5951        
 [3,]  2  49  2   1   2        1              NA                2096        
 [4,]  3  45  2   2   1        1               1                7882        
 [5,]  4  53  2   2   1        1               1                4870        
 [6,]  5  35  2   1   1       NA              NA                9055        
 [7,]  6  53  2   2   2        3              NA                2835        
 [8,]  7  35  2   3   3        1               2                6948        
 [9,]  8  61  2   1   2        4              NA                3059        
[10,]  9  28  2   3   2        1               2                5234        
[11,] 10  25  1   2   3        1               2                1295        
[12,] 11  24  1   2   3        1               1                4308        
[13,] 12  22  1   2   2        1               2                1567        
[14,] 13  60  2   1   2        1               1                1199        
[15,] 14  28  1   2   3        1               1                1403        
[16,] 15  32  1   1   2        2               1                1282        
[17,] 16  53  2   2   2       NA              NA                2424        
[18,] 17  25  2   2   2       NA               1                8072        
[19,] 18  44  1   3   1        1               2               12579        
[20,] 19  31  2   2   2        3              NA                3430        
[21,] 20  48  2   2   2        1              NA                2134        
[22,] 21  44  2   2   3        3               1                2647        
[23,] 22  48  2   1   3        1               1                2241        
[24,] 23  44  2   2   2        2               2                1804        
[25,] 24  26  2   2   2       NA              NA                2069        
[26,] 25  36  2   1   2        1               1                1374        
[27,] 26  39  2   1   2        1              NA                 426        
[28,] 27  42  1   2   3        4               3                 409        
[29,] 28  34  2   2   2        1               2                2415        
[30,] 29  63  2   2   2        1               1                6836        
[31,] ⋮   ⋮   ⋮   ⋮   ⋮       ⋮               ⋮                ⋮            
[32,] 970 22  2   2   2        2               2               1514         
[33,] 971 43  2   1   2        1              NA               7393         
[34,] 972 29  1   0   3        1               1               1193         
[35,] 973 36  2   2   3        1               1               7297         
[36,] 974 33  1   2   2        1              NA               2831         
[37,] 975 57  1   1   2        3               3               1258         
[38,] 976 64  1   2   2        1               2                753         
[39,] 977 42  2   2   2       NA               2               2427         
[40,] 978 47  2   1   2        1              NA               2538         
[41,] 979 25  2   2   3        2               2               1264         
[42,] 980 49  2   2   2        1               2               8386         
[43,] 981 33  2   3   3        1              NA               4844         
[44,] 982 28  1   3   2        2               3               2923         
[45,] 983 26  2   2   2        1               1               8229         
[46,] 984 30  2   1   2        1              NA               2028         
[47,] 985 25  1   2   3        1               1               1433         
[48,] 986 33  2   2   2        1               3               6289         
[49,] 987 64  1   2   2        2              NA               1409         
[50,] 988 29  2   3   1        1               1               6579         
[51,] 989 48  2   1   2        1               2               1743         
[5

In [170]:
german_data <- german_data
german_data[is.na(german_data <- german_data)] <- 0
german_data


X.  Age Sex Job Housing Saving.accounts Checking.account Credit.amount
 [1,]  0  67  2   2   2       0               1                 1169        
 [2,]  1  22  1   2   2       1               2                 5951        
 [3,]  2  49  2   1   2       1               0                 2096        
 [4,]  3  45  2   2   1       1               1                 7882        
 [5,]  4  53  2   2   1       1               1                 4870        
 [6,]  5  35  2   1   1       0               0                 9055        
 [7,]  6  53  2   2   2       3               0                 2835        
 [8,]  7  35  2   3   3       1               2                 6948        
 [9,]  8  61  2   1   2       4               0                 3059        
[10,]  9  28  2   3   2       1               2                 5234        
[11,] 10  25  1   2   3       1               2                 1295        
[12,] 11  24  1   2   3       1               1                 4308        
[13,] 12  22  1   2   2       1               2                 1567        
[14,] 13  60  2   1   2       1               1                 1199        
[15,] 14  28  1   2   3       1               1                 1403        
[16,] 15  32  1   1   2       2               1                 1282        
[17,] 16  53  2   2   2       0               0                 2424        
[18,] 17  25  2   2   2       0               1                 8072        
[19,] 18  44  1   3   1       1               2                12579        
[20,] 19  31  2   2   2       3               0                 3430        
[21,] 20  48  2   2   2       1               0                 2134        
[22,] 21  44  2   2   3       3               1                 2647        
[23,] 22  48  2   1   3       1               1                 2241        
[24,] 23  44  2   2   2       2               2                 1804        
[25,] 24  26  2   2   2       0               0                 2069        
[26,] 25  36  2   1   2       1               1                 1374        
[27,] 26  39  2   1   2       1               0                  426        
[28,] 27  42  1   2   3       4               3                  409        
[29,] 28  34  2   2   2       1               2                 2415        
[30,] 29  63  2   2   2       1               1                 6836        
[31,] ⋮   ⋮   ⋮   ⋮   ⋮       ⋮               ⋮                ⋮            
[32,] 970 22  2   2   2       2               2                1514         
[33,] 971 43  2   1   2       1               0                7393         
[34,] 972 29  1   0   3       1               1                1193         
[35,] 973 36  2   2   3       1               1                7297         
[36,] 974 33  1   2   2       1               0                2831         
[37,] 975 57  1   1   2       3               3                1258         
[38,] 976 64  1   2   2       1               2                 753         
[39,] 977 42  2   2   2       0               2                2427         
[40,] 978 47  2   1   2       1               0                2538         
[41,] 979 25  2   2   3       2               2                1264         
[42,] 980 49  2   2   2       1               2                8386         
[43,] 981 33  2   3   3       1               0                4844         
[44,] 982 28  1   3   2       2               3                2923         
[45,] 983 26  2   2   2       1               1                8229         
[46,] 984 30  2   1   2       1               0                2028         
[47,] 985 25  1   2   3       1               1                1433         
[48,] 986 33  2   2   2       1               3                6289         
[49,] 987 64  1   2   2       2               0                1409         
[50,] 988 29  2   3   1       1               1                6579         
[51,] 989 48  2   1   2       1               2                1743         
[5

In [155]:
summary(german_data)

       X.             Age             Sex            Job           Housing     
 Min.   :  0.0   Min.   :19.00   Min.   :1.00   Min.   :0.000   Min.   :1.000  
 1st Qu.:249.8   1st Qu.:27.00   1st Qu.:1.00   1st Qu.:2.000   1st Qu.:2.000  
 Median :499.5   Median :33.00   Median :2.00   Median :2.000   Median :2.000  
 Mean   :499.5   Mean   :35.55   Mean   :1.69   Mean   :1.904   Mean   :2.071  
 3rd Qu.:749.2   3rd Qu.:42.00   3rd Qu.:2.00   3rd Qu.:2.000   3rd Qu.:2.000  
 Max.   :999.0   Max.   :75.00   Max.   :2.00   Max.   :3.000   Max.   :3.000  
 Saving.accounts Checking.account Credit.amount      Duration   
 Min.   :0.00    Min.   :0.000    Min.   :  250   Min.   : 4.0  
 1st Qu.:1.00    1st Qu.:0.000    1st Qu.: 1366   1st Qu.:12.0  
 Median :1.00    Median :1.000    Median : 2320   Median :18.0  
 Mean   :1.19    Mean   :1.001    Mean   : 3271   Mean   :20.9  
 3rd Qu.:1.00    3rd Qu.:2.000    3rd Qu.: 3972   3rd Qu.:24.0  
 Max.   :4.00    Max.   :3.000    Max.   :18424   

In [171]:
normalize <- function(x) {
    return ((x - min(x)) / (max(x) - min(x)))
  }

In [172]:
german_data<-as.data.frame(german_data)
german_data_normed <- as.data.frame(lapply(german_data, normalize))
german_data_normed

X.          Age        Sex Job       Housing Saving.accounts
1    0.000000000 0.85714286 1   0.6666667 0.5     0.00           
2    0.001001001 0.05357143 0   0.6666667 0.5     0.25           
3    0.002002002 0.53571429 1   0.3333333 0.5     0.25           
4    0.003003003 0.46428571 1   0.6666667 0.0     0.25           
5    0.004004004 0.60714286 1   0.6666667 0.0     0.25           
6    0.005005005 0.28571429 1   0.3333333 0.0     0.00           
7    0.006006006 0.60714286 1   0.6666667 0.5     0.75           
8    0.007007007 0.28571429 1   1.0000000 1.0     0.25           
9    0.008008008 0.75000000 1   0.3333333 0.5     1.00           
10   0.009009009 0.16071429 1   1.0000000 0.5     0.25           
11   0.010010010 0.10714286 0   0.6666667 1.0     0.25           
12   0.011011011 0.08928571 0   0.6666667 1.0     0.25           
13   0.012012012 0.05357143 0   0.6666667 0.5     0.25           
14   0.013013013 0.73214286 1   0.3333333 0.5     0.25           
15   0.014014014 0.16071429 0   0.6666667 1.0     0.25           
16   0.015015015 0.23214286 0   0.3333333 0.5     0.50           
17   0.016016016 0.60714286 1   0.6666667 0.5     0.00           
18   0.017017017 0.10714286 1   0.6666667 0.5     0.00           
19   0.018018018 0.44642857 0   1.0000000 0.0     0.25           
20   0.019019019 0.21428571 1   0.6666667 0.5     0.75           
21   0.020020020 0.51785714 1   0.6666667 0.5     0.25           
22   0.021021021 0.44642857 1   0.6666667 1.0     0.75           
23   0.022022022 0.51785714 1   0.3333333 1.0     0.25           
24   0.023023023 0.44642857 1   0.6666667 0.5     0.50           
25   0.024024024 0.12500000 1   0.6666667 0.5     0.00           
26   0.025025025 0.30357143 1   0.3333333 0.5     0.25           
27   0.026026026 0.35714286 1   0.3333333 0.5     0.25           
28   0.027027027 0.41071429 0   0.6666667 1.0     1.00           
29   0.028028028 0.26785714 1   0.6666667 0.5     0.25           
30   0.029029029 0.78571429 1   0.6666667 0.5     0.25           
⋮    ⋮           ⋮          ⋮   ⋮         ⋮       ⋮              
971  0.970971    0.05357143 1   0.6666667 0.5     0.50           
972  0.971972    0.42857143 1   0.3333333 0.5     0.25           
973  0.972973    0.17857143 0   0.0000000 1.0     0.25           
974  0.973974    0.30357143 1   0.6666667 1.0     0.25           
975  0.974975    0.25000000 0   0.6666667 0.5     0.25           
976  0.975976    0.67857143 0   0.3333333 0.5     0.75           
977  0.976977    0.80357143 0   0.6666667 0.5     0.25           
978  0.977978    0.41071429 1   0.6666667 0.5     0.00           
979  0.978979    0.50000000 1   0.3333333 0.5     0.25           
980  0.979980    0.10714286 1   0.6666667 1.0     0.50           
981  0.980981    0.53571429 1   0.6666667 0.5     0.25           
982  0.981982    0.25000000 1   1.0000000 1.0     0.25           
983  0.982983    0.16071429 0   1.0000000 0.5     0.50           
984  0.983984    0.12500000 1   0.6666667 0.5     0.25           
985  0.984985    0.19642857 1   0.3333333 0.5     0.25           
986  0.985986    0.10714286 0   0.6666667 1.0     0.25           
987  0.986987    0.25000000 1   0.6666667 0.5     0.25           
988  0.987988    0.80357143 0   0.6666667 0.5     0.50           
989  0.988989    0.17857143 1   1.0000000 0.0     0.25           
990  0.989990    0.51785714 1   0.3333333 0.5     0.25           
991  0.990991    0.32142857 1   0.3333333 0.5     0.00           
992  0.991992    0.26785714 1   0.3333333 0.5     0.50           
993  0.992993    0.07142857 1   0.3333333 1.0     0.00           
994  0.993994    0.19642857 1   1.0000000 0.5     0.25           
995  0.994995    0.55357143 1   0.6666667 0.5     0.00           
996  0.995996    0.21428571 0   0.3333333 0.5     0.25           
997  0.996997    0.37500000 1   1.0000000 0.5     0.25           
998  0.997998    0.33928571 1   0.6666667 0.5     0.25           
999  0.998999    0.07142857 1   0.6666667 0.0

In [173]:
str(german_data_normed)

'data.frame':	1000 obs. of  11 variables:
 $ X.              : num  0 0.001 0.002 0.003 0.004 ...
 $ Age             : num  0.8571 0.0536 0.5357 0.4643 0.6071 ...
 $ Sex             : num  1 0 1 1 1 1 1 1 1 1 ...
 $ Job             : num  0.667 0.667 0.333 0.667 0.667 ...
 $ Housing         : num  0.5 0.5 0.5 0 0 0 0.5 1 0.5 0.5 ...
 $ Saving.accounts : num  0 0.25 0.25 0.25 0.25 0 0.75 0.25 1 0.25 ...
 $ Checking.account: num  0.333 0.667 0 0.333 0.333 ...
 $ Credit.amount   : num  0.0506 0.3137 0.1016 0.4199 0.2542 ...
 $ Duration        : num  0.0294 0.6471 0.1176 0.5588 0.2941 ...
 $ Purpose         : num  0.714 0.714 0.429 0.571 0.143 ...
 $ Credit.Risks    : num  0 1 0 0 1 0 0 0 0 1 ...


In [174]:
german_data_normed<-as.data.frame(german_data_normed)
german_data_normed$Credit.Risks<-factor(german_data_normed$Credit.Risks, label=c("low","high"))
#german_data_subset=data.frame(german_data_subset)
head(german_data_normed)

X.          Age        Sex Job       Housing Saving.accounts Checking.account
1 0.000000000 0.85714286 1   0.6666667 0.5     0.00            0.3333333       
2 0.001001001 0.05357143 0   0.6666667 0.5     0.25            0.6666667       
3 0.002002002 0.53571429 1   0.3333333 0.5     0.25            0.0000000       
4 0.003003003 0.46428571 1   0.6666667 0.0     0.25            0.3333333       
5 0.004004004 0.60714286 1   0.6666667 0.0     0.25            0.3333333       
6 0.005005005 0.28571429 1   0.3333333 0.0     0.00            0.0000000       
  Credit.amount Duration   Purpose   Credit.Risks
1 0.05056674    0.02941176 0.7142857 low         
2 0.31368989    0.64705882 0.7142857 high        
3 0.10157368    0.11764706 0.4285714 low         
4 0.41994057    0.55882353 0.5714286 low         
5 0.25420931    0.29411765 0.1428571 high        
6 0.48448333    0.47058824 0.4285714 low

In [176]:
n.points=nrow(german_data_normed)
n.points

[1] 1000

<b>80% Training Set ; 20% Test Set<b>

In [281]:
set.seed(125)
sampling_rate_1 <-0.8
num.test.set.labels<- n.points* (1 - sampling_rate_1)

training<-sample(1:n.points, sampling_rate_1*n.points, replace=FALSE)
train_1<- subset(german_data_normed[training,],select=c(Age,Sex,Job,Housing,Saving.accounts,Checking.account,Credit.amount,Duration,Purpose))
#train
testing<-setdiff(1:n.points,training)
test_1<-subset(german_data_normed[testing,], select=c(Age,Sex,Job,Housing,Saving.accounts,Checking.account,Credit.amount,Duration,Purpose))
#test
cl_1<- german_data_normed$Credit.Risks[training]
#cl
true.labels<-german_data_normed$Credit.Risks[testing]
#true.labels


In [282]:
dim(train_1)
dim(test_1)
length(cl_1)


[1] 800   9

[1] 200   9

[1] 800

In [283]:
knn(train_1,test_1,cl_1,k=3) 

[1] low  low  low  low  low  high low  low  low  low  low  low  low  low  low 
 [16] low  high low  low  high high high low  low  low  low  low  low  low  high
 [31] low  low  low  low  high low  low  low  low  low  low  high low  low  low 
 [46] low  high low  low  low  low  low  low  high low  low  high low  high high
 [61] low  low  low  high low  low  low  low  high low  low  low  high low  high
 [76] high low  low  low  low  high low  low  low  low  high high high low  high
 [91] high low  high low  low  low  high low  low  low  low  low  low  low  high
[106] high low  high high low  high low  high high low  low  low  low  high low 
[121] low  low  low  low  low  low  low  low  high low  high high high low  low 
[136] low  low  low  low  low  low  low  low  low  low  low  high high low  low 
[151] low  high high high low  low  low  low  low  high low  low  low  low  low 
[166] low  low  low  low  low  low  high low  high low  low  high high low  high
[181] low  low  low  high low  low  low  low  high low  low  low  low  low  high
[196] low  high low  low  high
Levels: low high

In [284]:
#compute and print misclassification rate
for(k_1 in 1:20) {
  print(k_1)
  predicted.labels <- knn(train_1,test_1,cl_1,k_1)
  num.incorrect.labels<-sum(predicted.labels !=true.labels)
  misclassification.rate_1<-num.incorrect.labels/num.test.set.labels
  print(misclassification.rate_1)
    
}



[1] 1
[1] 0.345
[1] 2
[1] 0.325
[1] 3
[1] 0.32
[1] 4
[1] 0.335
[1] 5
[1] 0.27
[1] 6
[1] 0.255
[1] 7
[1] 0.255
[1] 8
[1] 0.28
[1] 9
[1] 0.29
[1] 10
[1] 0.295
[1] 11
[1] 0.29
[1] 12
[1] 0.28
[1] 13
[1] 0.305
[1] 14
[1] 0.31
[1] 15
[1] 0.315
[1] 16
[1] 0.29
[1] 17
[1] 0.285
[1] 18
[1] 0.275
[1] 19
[1] 0.285
[1] 20
[1] 0.285


The lowest Misclassification Rate when the training data is 80% and test data is 20% is  at <b>k=6</b> being equal to <b>0.255</b>

In [285]:
knn(train_1,test_1,cl_1,k=6)

[1] high low  high low  low  high low  low  low  low  low  low  low  low  low 
 [16] low  low  low  low  high high low  low  low  low  low  low  low  low  low 
 [31] high low  low  low  low  low  low  high low  low  low  high low  low  low 
 [46] low  high low  low  low  low  low  high high low  low  high low  high high
 [61] low  low  low  low  low  low  low  low  low  high low  low  high low  high
 [76] high low  low  high low  low  low  low  low  low  high high low  low  high
 [91] low  low  low  low  low  low  high low  low  low  high low  high low  low 
[106] high low  low  high low  high low  high high low  low  low  low  high low 
[121] low  low  low  low  low  low  low  low  high low  high low  low  low  low 
[136] low  low  low  low  low  high low  high low  low  low  high high low  low 
[151] low  high high low  low  low  low  low  high high low  low  low  low  high
[166] low  low  low  low  low  low  low  low  low  low  low  high high low  high
[181] low  low  low  low  low  low  low  low  high low  low  low  low  low  high
[196] low  high low  low  high
Levels: low high

<b>90% Training Set ; 10% Test Set<b>

In [288]:
# 90% Training Set ; 10% Test Set
set.seed(125)
sampling_rate_2<-0.9
num.test.set.labels<- n.points* (1 - sampling_rate_2)

training<-sample(1:n.points, sampling_rate_2*n.points, replace=FALSE)
train_2<- subset(german_data_normed[training,],select=c(Age,Sex,Job,Housing,Saving.accounts,Checking.account,Credit.amount,Duration,Purpose))
#train
testing<-setdiff(1:n.points,training)
test_2<-subset(german_data_normed[testing,], select=c(Age,Sex,Job,Housing,Saving.accounts,Checking.account,Credit.amount,Duration,Purpose))
#test
cl_2<- german_data_normed$Credit.Risks[training]
#cl
true.labels<-german_data_normed$Credit.Risks[testing]
#true.labels
knn(train_2,test_2,cl_2,k=3) 

[1] low  low  high low  low  low  low  low  high high low  low  low  low  high
 [16] low  low  low  low  low  low  low  low  low  low  high low  low  high high
 [31] low  low  high low  low  low  high low  high low  low  low  low  high low 
 [46] low  low  low  low  low  low  high low  high low  low  low  low  low  low 
 [61] low  low  high high high low  low  low  low  low  low  high high low  high
 [76] high low  high low  low  low  low  low  low  high low  high low  low  high
 [91] high high low  low  high low  high low  high high
Levels: low high

In [289]:
#compute and print misclassification rate
for(k_2 in 1:20) {
  print(k_2)
  predicted.labels <- knn(train_2,test_2,cl_2,k_2)
  num.incorrect.labels<-sum(predicted.labels !=true.labels)
  misclassification.rate_2<-num.incorrect.labels/num.test.set.labels
  print(misclassification.rate_2)
}


[1] 1
[1] 0.28
[1] 2
[1] 0.26
[1] 3
[1] 0.31
[1] 4
[1] 0.33
[1] 5
[1] 0.31
[1] 6
[1] 0.33
[1] 7
[1] 0.27
[1] 8
[1] 0.28
[1] 9
[1] 0.27
[1] 10
[1] 0.25
[1] 11
[1] 0.25
[1] 12
[1] 0.29
[1] 13
[1] 0.26
[1] 14
[1] 0.28
[1] 15
[1] 0.28
[1] 16
[1] 0.28
[1] 17
[1] 0.3
[1] 18
[1] 0.3
[1] 19
[1] 0.29
[1] 20
[1] 0.29


The lowest Misclassification Rate when the training data is 90% and test data is 10% is at <b>k=10</b> being equal to <b>0.25</b>

In [290]:
knn(train_2,test_2,cl_2,k=10)

[1] low  low  high low  low  low  low  low  low  high low  low  low  low  low 
 [16] low  low  low  low  high low  low  low  low  high high low  low  high low 
 [31] low  low  low  low  low  low  high low  high low  low  low  low  high low 
 [46] low  low  high low  low  low  low  low  high low  high low  low  low  low 
 [61] low  low  high low  low  low  low  high high low  low  high high low  high
 [76] high low  high low  low  low  low  low  low  low  high low  low  low  low 
 [91] high low  low  low  low  low  high low  low  low 
Levels: low high

<b>75% Training Set ; 25% Test Set<b>

In [338]:
set.seed(150)
sampling_rate_3<-0.75
num.test.set.labels<- n.points* (1 - sampling_rate_3)

training<-sample(1:n.points, sampling_rate_3*n.points, replace=FALSE)
train_3<- subset(german_data_normed[training,],select=c(Age,Sex,Job,Housing,Saving.accounts,Checking.account,Credit.amount,Duration,Purpose))
#train
testing<-setdiff(1:n.points,training)
test_3<-subset(german_data_normed[testing,], select=c(Age,Sex,Job,Housing,Saving.accounts,Checking.account,Credit.amount,Duration,Purpose))
#test
cl_3<- german_data_normed$Credit.Risks[training]
#cl
true.labels<-german_data_normed$Credit.Risks[testing]
#true.labels
knn(train_3,test_3,cl_3,k=3) 

[1] low  low  high low  high low  low  low  low  low  low  low  low  low  low 
 [16] low  high low  low  low  low  low  low  high low  low  low  low  low  high
 [31] low  low  low  low  low  low  low  high low  low  low  low  high high low 
 [46] low  low  low  low  low  high high high low  low  low  low  low  high low 
 [61] low  high high low  low  low  low  low  low  low  low  low  low  low  high
 [76] low  low  low  high high low  low  low  low  low  low  high high low  high
 [91] low  low  low  low  low  low  low  high low  high low  low  high low  low 
[106] low  low  low  low  low  high low  low  low  low  low  low  low  low  high
[121] low  low  low  low  low  low  low  low  high low  low  low  low  low  high
[136] high low  low  low  high low  low  high low  low  high high low  high low 
[151] low  low  high low  low  high high high high low  low  high low  low  low 
[166] high high high low  low  low  low  high low  low  low  low  low  low  low 
[181] low  high high low  low  high low  high low  high high low  low  low  low 
[196] low  low  high low  low  low  low  low  low  low  low  low  high low  low 
[211] low  low  low  low  high low  low  low  high low  low  high low  high low 
[226] low  low  high low  high low  low  low  low  low  low  low  high low  low 
[241] low  low  low  low  low  low  low  low  low  low 
Levels: low high

In [339]:
#compute and print misclassification rate
for(k_3 in 1:20) {
  print(k_3)
  predicted.labels <- knn(train_3,test_3,cl_3,k_3)
  num.incorrect.labels<-sum(predicted.labels !=true.labels)
  misclassification.rate_3<-num.incorrect.labels/num.test.set.labels
  print(misclassification.rate_3)
}

[1] 1
[1] 0.272
[1] 2
[1] 0.292
[1] 3
[1] 0.264
[1] 4
[1] 0.292
[1] 5
[1] 0.256
[1] 6
[1] 0.276
[1] 7
[1] 0.288
[1] 8
[1] 0.288
[1] 9
[1] 0.284
[1] 10
[1] 0.296
[1] 11
[1] 0.284
[1] 12
[1] 0.264
[1] 13
[1] 0.268
[1] 14
[1] 0.256
[1] 15
[1] 0.268
[1] 16
[1] 0.252
[1] 17
[1] 0.268
[1] 18
[1] 0.256
[1] 19
[1] 0.264
[1] 20
[1] 0.26


The lowest Misclassification Rate when the training data is 75% and test data is 25% is at <b>k=16</b> being equal to <b>0.252</b>

In [340]:
knn(train_3,test_3,cl_3,k=16)

[1] low  low  high low  low  low  low  low  low  low  low  low  low  low  high
 [16] low  high low  low  low  low  low  low  low  low  low  low  low  low  low 
 [31] low  low  low  low  high low  low  high low  low  low  low  low  low  low 
 [46] low  low  low  low  low  low  low  low  low  low  low  low  low  low  low 
 [61] low  low  high low  low  low  high low  low  low  low  low  low  low  high
 [76] low  low  low  high high low  low  low  low  low  low  low  high low  low 
 [91] low  low  low  low  low  low  low  low  low  low  high high low  low  low 
[106] low  low  high low  low  low  low  low  low  low  low  low  low  low  low 
[121] low  low  low  low  low  low  low  low  high low  low  high low  low  low 
[136] low  high low  low  high low  low  low  low  low  low  low  low  low  low 
[151] high low  high low  high low  low  low  low  low  low  low  low  low  low 
[166] high high low  low  low  low  low  low  low  low  low  low  low  low  low 
[181] low  high high low  low  low  low  low  low  high high high low  low  high
[196] low  low  high low  low  low  low  low  low  high low  low  low  low  low 
[211] low  low  low  low  low  low  low  low  low  low  low  low  low  low  low 
[226] low  low  low  low  high low  low  low  low  low  low  low  low  low  low 
[241] low  low  low  low  low  low  low  low  low  low 
Levels: low high

<b>70% Training Set ; 30% Test Set<b>

In [309]:
set.seed(125)
sampling_rate_4<-0.70
num.test.set.labels<- n.points* (1 - sampling_rate_4)

training<-sample(1:n.points, sampling_rate_4*n.points, replace=FALSE)
train_4<- subset(german_data_normed[training,],select=c(Age,Sex,Job,Housing,Saving.accounts,Checking.account,Credit.amount,Duration,Purpose))
#train
testing<-setdiff(1:n.points,training)
test_4<-subset(german_data_normed[testing,], select=c(Age,Sex,Job,Housing,Saving.accounts,Checking.account,Credit.amount,Duration,Purpose))
#test
cl_4<- german_data_normed$Credit.Risks[training]
#cl
true.labels<-german_data_normed$Credit.Risks[testing]
#true.labels
knn(train_4,test_4,cl_4,k=3) 

[1] low  low  low  low  low  low  low  high low  low  low  low  low  low  low 
 [16] low  low  low  low  low  low  low  low  low  low  high low  low  low  high
 [31] high high high low  low  low  low  low  low  low  low  high low  low  low 
 [46] low  low  high low  low  low  low  low  high low  low  low  high high low 
 [61] high low  low  low  low  low  low  low  high low  low  low  low  low  low 
 [76] low  high low  low  low  high high low  high high high low  low  low  low 
 [91] high low  low  low  low  low  low  low  high low  low  low  low  high high
[106] low  low  high high low  high low  low  high low  low  low  low  low  high
[121] high low  low  low  low  low  high high low  high high low  high high low 
[136] low  high low  low  low  low  low  high low  low  low  low  low  low  low 
[151] low  low  low  low  low  high low  high high low  low  high high low  low 
[166] high low  high low  high high high low  low  low  low  low  high low  high
[181] low  low  low  low  low  low  low  low  low  low  low  high low  high low 
[196] low  high low  high low  low  high low  low  low  low  low  low  low  low 
[211] low  low  low  low  low  high low  high low  low  high high low  low  low 
[226] low  low  high low  high high low  high low  low  low  low  low  low  low 
[241] high low  low  high low  high high low  low  low  low  low  low  low  low 
[256] low  low  low  low  low  low  low  high high high low  high low  low  high
[271] high low  high low  low  low  low  high high low  low  low  low  low  high
[286] low  low  low  low  low  low  high high low  high low  low  low  high high
Levels: low high

In [310]:
#compute and print misclassification rate
for(k_4 in 1:20) {
  print(k_4)
  predicted.labels <- knn(train_4,test_4,cl_4,k_4)
  num.incorrect.labels<-sum(predicted.labels !=true.labels)
  misclassification.rate_4<-num.incorrect.labels/num.test.set.labels
  print(misclassification.rate_4)
}

[1] 1
[1] 0.34
[1] 2
[1] 0.3533333
[1] 3
[1] 0.32
[1] 4
[1] 0.3133333
[1] 5
[1] 0.2966667
[1] 6
[1] 0.2966667
[1] 7
[1] 0.3033333
[1] 8
[1] 0.3233333
[1] 9
[1] 0.3033333
[1] 10
[1] 0.33
[1] 11
[1] 0.3033333
[1] 12
[1] 0.3033333
[1] 13
[1] 0.3166667
[1] 14
[1] 0.32
[1] 15
[1] 0.3
[1] 16
[1] 0.3
[1] 17
[1] 0.2966667
[1] 18
[1] 0.2933333
[1] 19
[1] 0.2966667
[1] 20
[1] 0.3066667


The lowest Misclassification Rate when the training data is 70% and test data is 30% is at <b>k=18</b> being equal to <b>0.2933333</b>

In [311]:
knn(train_4,test_4,cl_4,k=18)

[1] high high low  low  low  low  low  low  low  low  low  low  low  low  low 
 [16] low  low  low  low  low  low  low  low  low  low  low  low  low  low  low 
 [31] low  low  low  low  low  low  low  low  low  low  low  low  low  low  low 
 [46] low  low  low  low  low  low  low  low  high low  high low  high low  low 
 [61] low  low  low  low  low  low  low  low  low  high low  low  low  low  low 
 [76] high low  low  low  low  high low  low  low  low  low  high low  low  low 
 [91] low  low  low  high low  low  low  low  high low  low  low  low  low  high
[106] high low  high high low  high low  low  low  low  low  low  low  low  low 
[121] low  low  low  low  low  low  low  low  low  high low  low  low  high low 
[136] low  low  low  low  low  low  low  high low  low  low  low  low  low  low 
[151] low  low  low  low  low  low  low  low  low  low  low  low  high low  low 
[166] low  high high low  low  high low  low  low  low  low  low  low  low  low 
[181] low  low  high low  low  low  low  low  low  low  low  low  low  low  low 
[196] low  low  low  low  low  low  low  low  low  low  low  low  low  low  high
[211] low  low  low  high low  high low  low  low  low  high low  low  low  low 
[226] low  low  high low  low  high high high low  low  low  low  low  low  high
[241] low  high low  high low  high low  low  low  high low  low  low  low  low 
[256] low  low  low  low  low  low  low  low  low  low  low  low  low  low  low 
[271] high low  low  low  low  low  low  low  low  low  low  low  low  low  high
[286] low  low  low  low  low  low  low  low  low  low  low  low  high low  high
Levels: low high

<b>60% Training Set ; 40% Test Set<b>

In [341]:
set.seed(150)
sampling_rate_5<-0.60
num.test.set.labels<- n.points* (1 - sampling_rate_5)

training<-sample(1:n.points, sampling_rate_5*n.points, replace=FALSE)
train_5<- subset(german_data_normed[training,],select=c(Age,Sex,Job,Housing,Saving.accounts,Checking.account,Credit.amount,Duration,Purpose))
#train
testing<-setdiff(1:n.points,training)
test_5<-subset(german_data_normed[testing,], select=c(Age,Sex,Job,Housing,Saving.accounts,Checking.account,Credit.amount,Duration,Purpose))
#test
cl_5<- german_data_normed$Credit.Risks[training]
#cl
true.labels<-german_data_normed$Credit.Risks[testing]
#true.labels
knn(train_5,test_5,cl_5,k=3) 

[1] high low  low  high low  low  low  low  high low  low  high high low  low 
 [16] low  high low  low  low  low  low  low  low  low  low  high low  low  low 
 [31] low  low  low  high high low  low  high high low  low  low  low  high low 
 [46] low  low  low  low  high high low  low  low  low  high low  low  low  low 
 [61] low  low  low  low  low  high low  low  high low  low  low  low  low  low 
 [76] low  low  high low  high low  high low  high low  low  low  low  low  low 
 [91] low  high low  low  high high low  low  low  low  high low  low  low  low 
[106] low  low  low  low  low  low  high low  low  low  low  low  high high high
[121] low  low  low  low  low  low  low  low  high low  low  low  high high high
[136] low  low  high high low  low  high low  high high low  low  high low  low 
[151] low  high low  low  low  low  low  low  high low  low  high low  low  low 
[166] low  low  low  low  low  low  low  low  low  low  low  low  low  low  low 
[181] low  low  low  low  low  low  low  low  low  low  low  high low  low  low 
[196] low  low  low  low  low  low  low  high low  low  low  low  low  low  low 
[211] low  low  low  high high low  high low  low  low  low  high low  low  high
[226] low  high low  high low  low  high high low  low  low  low  low  low  high
[241] low  low  low  high high low  high low  high low  high low  high low  low 
[256] low  low  low  low  high low  high low  high high high low  low  low  low 
[271] low  low  high high low  low  low  low  high high low  low  low  high low 
[286] low  low  low  low  low  low  low  low  high high low  high low  low  high
[301] high low  high high low  high low  low  low  high high high high low  low 
[316] high low  high low  high high low  low  low  low  high low  high low  high
[331] low  low  low  low  low  low  low  low  high low  low  low  low  low  low 
[346] low  low  low  high low  low  high low  low  high low  low  low  low  high
[361] low  high low  low  low  high low  high low  low  high low  low  low  low 
[376] low  low  low  high low  low  high low  low  low  low  low  low  low  low 
[391] low  low  low  low  low  low  high low  low  low 
Levels: low high

In [342]:
#compute and print misclassification rate
for(k_5 in 1:20) {
  print(k_5)
  predicted.labels <- knn(train_5,test_5,cl_5,k_5)
  num.incorrect.labels<-sum(predicted.labels !=true.labels)
  misclassification.rate_5<-num.incorrect.labels/num.test.set.labels
  print(misclassification.rate_5)
}

[1] 1
[1] 0.3275
[1] 2
[1] 0.3575
[1] 3
[1] 0.32
[1] 4
[1] 0.335
[1] 5
[1] 0.315
[1] 6
[1] 0.3175
[1] 7
[1] 0.3225
[1] 8
[1] 0.305
[1] 9
[1] 0.3275
[1] 10
[1] 0.315
[1] 11
[1] 0.2925
[1] 12
[1] 0.295
[1] 13
[1] 0.29
[1] 14
[1] 0.2875
[1] 15
[1] 0.265
[1] 16
[1] 0.2725
[1] 17
[1] 0.2775
[1] 18
[1] 0.2775
[1] 19
[1] 0.275
[1] 20
[1] 0.28


The lowest Misclassification Rate when the training data is 60% and test data is 40% is at <b>k=15</b> being equal to <b>0.265</b>

In [343]:
knn(train_5,test_5,cl_5,k=15)

[1] low  low  low  low  low  low  high low  low  low  low  low  low  low  low 
 [16] low  low  low  low  low  low  low  low  low  low  low  low  low  low  low 
 [31] low  low  low  high low  low  low  low  high low  low  low  low  low  low 
 [46] low  low  low  low  low  high high low  low  low  high low  low  low  low 
 [61] low  low  low  low  low  low  low  low  low  low  low  low  low  low  low 
 [76] low  low  low  low  high low  low  low  high low  low  low  low  low  low 
 [91] low  high low  low  high low  low  low  low  low  high low  low  low  high
[106] low  low  low  low  low  low  high low  low  low  low  low  high high low 
[121] low  low  low  low  low  low  low  low  low  low  low  low  low  low  high
[136] low  low  low  low  low  low  low  low  low  high low  low  low  low  low 
[151] high low  low  low  low  high low  low  high low  low  low  high low  low 
[166] low  low  low  high low  high low  low  low  low  low  low  low  low  low 
[181] low  low  low  low  low  low  low  low  low  high low  low  low  low  low 
[196] low  low  low  low  low  low  low  high low  low  low  high low  low  high
[211] low  low  low  high low  low  high low  low  low  low  low  low  low  high
[226] low  low  low  high low  high low  low  low  low  low  high low  low  high
[241] low  low  low  low  high low  low  low  low  low  high low  low  low  low 
[256] high low  low  low  high low  high low  low  high high low  low  low  low 
[271] low  low  high low  low  low  high low  high low  low  low  low  high low 
[286] low  low  low  high low  low  low  low  high high low  low  low  low  low 
[301] high low  low  low  low  low  low  low  low  high high high low  low  low 
[316] low  low  high high high high low  low  low  high low  low  high low  low 
[331] low  low  low  low  high low  low  low  low  low  low  low  low  low  low 
[346] low  low  low  low  low  low  low  low  low  low  low  high low  low  low 
[361] low  low  low  low  low  low  low  high low  low  low  low  low  low  low 
[376] low  low  low  low  low  low  low  low  low  low  low  low  low  high low 
[391] low  low  low  low  high low  low  low  low  low 
Levels: low high

<b>50% Training Set ; 50% Test Set<b>

In [345]:
set.seed(125)
sampling_rate_6<-0.50
num.test.set.labels<- n.points* (1 - sampling_rate_6)

training<-sample(1:n.points, sampling_rate_5*n.points, replace=FALSE)
train_6<- subset(german_data_normed[training,],select=c(Age,Sex,Job,Housing,Saving.accounts,Checking.account,Credit.amount,Duration,Purpose))
#train
testing<-setdiff(1:n.points,training)
test_6<-subset(german_data_normed[testing,], select=c(Age,Sex,Job,Housing,Saving.accounts,Checking.account,Credit.amount,Duration,Purpose))
#test
cl_6<- german_data_normed$Credit.Risks[training]
#cl
true.labels<-german_data_normed$Credit.Risks[testing]
#true.labels
knn(train_6,test_6,cl_6,k=3) 

[1] high low  low  low  low  high low  low  high low  low  low  high low  low 
 [16] low  low  low  low  low  low  low  low  low  high low  high low  low  low 
 [31] low  low  high low  low  low  low  high low  low  low  low  low  high high
 [46] low  high low  high high low  low  low  low  high low  low  low  low  low 
 [61] low  low  low  low  low  high low  low  high low  low  low  low  high low 
 [76] high low  low  high low  low  high low  low  low  low  high high low  high
 [91] low  low  low  low  low  low  low  low  high low  low  low  low  low  low 
[106] low  low  low  low  low  low  low  low  high high low  low  low  high low 
[121] low  low  low  low  low  low  low  high low  low  low  low  low  low  low 
[136] low  high low  low  low  high low  high high low  low  low  low  high low 
[151] high low  low  low  high high low  low  low  low  high low  low  low  low 
[166] low  low  low  low  high high low  high low  low  low  low  low  low  high
[181] high low  high high low  high low  low  low  low  high high low  low  low 
[196] low  low  low  low  low  low  low  low  low  low  low  low  high low  low 
[211] low  low  low  low  low  low  high low  low  high high low  low  low  low 
[226] high low  high high low  high low  low  high low  high low  high high high
[241] low  high low  low  low  low  low  high low  low  high low  low  low  low 
[256] low  low  low  low  high low  low  low  low  low  high low  high low  low 
[271] high low  high low  low  low  low  high low  low  low  low  high low  low 
[286] low  low  high low  low  low  low  low  low  high low  high low  low  high
[301] low  high high low  low  low  low  low  low  low  high low  high high low 
[316] high high low  low  low  low  low  low  low  low  high low  low  high low 
[331] high low  low  low  low  low  low  high low  low  low  low  low  low  low 
[346] low  low  low  low  low  low  low  low  high high low  low  low  high high
[361] low  high low  high high low  high low  low  low  low  high high low  low 
[376] low  low  low  low  high low  high low  low  low  low  low  low  low  high
[391] high low  high low  low  high high low  low  high
Levels: low high

In [346]:
#compute and print misclassification rate
for(k_6 in 1:20) {
  print(k_6)
  predicted.labels <- knn(train_6,test_6,cl_6,k_6)
  num.incorrect.labels<-sum(predicted.labels !=true.labels)
  misclassification.rate_6<-num.incorrect.labels/num.test.set.labels
  print(misclassification.rate_6)
}

[1] 1
[1] 0.28
[1] 2
[1] 0.282
[1] 3
[1] 0.264
[1] 4
[1] 0.25
[1] 5
[1] 0.252
[1] 6
[1] 0.272
[1] 7
[1] 0.262
[1] 8
[1] 0.262
[1] 9
[1] 0.258
[1] 10
[1] 0.278
[1] 11
[1] 0.26
[1] 12
[1] 0.254
[1] 13
[1] 0.244
[1] 14
[1] 0.246
[1] 15
[1] 0.25
[1] 16
[1] 0.244
[1] 17
[1] 0.244
[1] 18
[1] 0.234
[1] 19
[1] 0.238
[1] 20
[1] 0.24


The lowest Misclassification Rate when the training data is 50% and test data is 50% is at <b>k=18</b> being equal to <b>0.234</b>



In [347]:
knn(train_5,test_5,cl_5,k=18)

[1] low  low  low  low  low  low  high low  low  low  low  low  low  low  low 
 [16] low  low  low  low  low  low  low  low  low  low  low  low  low  low  low 
 [31] low  low  low  high high low  low  low  high low  low  low  low  low  low 
 [46] low  low  low  low  low  high high low  low  low  high low  low  low  low 
 [61] low  low  low  low  low  low  low  low  low  low  low  low  low  low  low 
 [76] low  high low  low  high low  low  low  high low  low  low  low  low  low 
 [91] low  high low  low  low  low  low  low  low  low  high low  low  low  high
[106] low  low  low  low  low  low  low  low  low  low  low  low  high low  low 
[121] low  low  low  low  low  low  low  low  low  low  low  low  high low  high
[136] low  low  low  low  low  low  low  low  low  high low  low  low  low  low 
[151] high low  low  low  low  high low  low  low  low  low  low  low  low  low 
[166] low  low  low  high low  high low  low  low  low  low  low  low  low  low 
[181] low  low  low  low  low  low  low  low  low  high low  low  low  high low 
[196] low  low  low  low  low  low  low  high low  low  low  high low  low  high
[211] low  low  low  high low  low  high low  low  low  low  low  low  low  high
[226] low  high low  high low  low  low  low  low  low  low  high low  low  high
[241] low  low  low  low  high low  low  low  low  low  high low  low  low  low 
[256] high low  low  low  high low  high low  low  high high low  low  low  low 
[271] low  low  high low  low  low  high low  high low  low  low  low  high low 
[286] low  low  low  high low  low  low  low  high high low  high low  low  high
[301] high low  high low  low  low  low  low  low  high high high low  low  low 
[316] low  low  high high high high low  low  low  high high low  high low  low 
[331] low  low  low  low  high low  low  low  low  low  low  low  low  low  low 
[346] low  low  low  low  low  low  low  low  low  low  low  low  low  low  high
[361] low  low  low  low  low  low  low  high low  low  low  low  low  low  low 
[376] low  low  low  low  low  low  low  low  low  low  low  low  low  high low 
[391] low  low  low  low  high low  low  low  low  low 
Levels: low high